dallinger/config.py

```diff
@@ -0,0 +1,163 @@
+from __future__ import unicode_literals
+
+from collections import deque
+from ConfigParser import SafeConfigParser
+import distutils.util
+import os
+import threading
+
+marker = object()
+
+try:
+    unicode = unicode
+except NameError:  # Python 3
+    unicode = str
+
+
+class Configuration(object):
+
+    SUPPORTED_TYPES = {
+        bytes,
+        unicode,
+        int,
+        float,
+        bool,
+    }
+
+    def __init__(self):
+        self.data = deque()
+        self.types = {}
+        self.synonyms = {}
+        self.ready = False
+
+    def extend(self, mapping, cast_types=False):
+        normalised_mapping = {}
+        for key, value in mapping.items():
+            key = self.synonyms.get(key, key)
+            if key not in self.types:
+                # This key hasn't been registered, we ignore it
+                continue
+            expected_type = self.types.get(key)
+            if cast_types:
+                try:
+                    if expected_type == bool:
+                        value = distutils.util.strtobool(value)
+                    value = expected_type(value)
+                except ValueError:
+                    pass
+            if not isinstance(value, expected_type):
+                raise ValueError(
+                    'Got {value} for {key}, expected {expected_type}'
+                    .format(
+                        value=repr(value),
+                        key=key,
+                        expected_type=expected_type,
+                    )
+                )
+            normalised_mapping[key] = value
+        self.data.extendleft([normalised_mapping])
+
+    def get(self, key, default=marker):
+        if not self.ready:
+            raise RuntimeError('Config loading not finished')
+        for layer in self.data:
+            try:
+                return layer[key]
+            except KeyError:
+                continue
+        if default is marker:
+            raise KeyError(key)
+        return default
+
+    def register(self, key, type_, synonyms=set()):
+        if type_ not in self.SUPPORTED_TYPES:
+            raise ValueError(
+                '{type} is not a supported type'.format(
+                    type=type_
+                )
+            )
+        self.types[key] = type_
+        for synonym in synonyms:
+            self.synonyms[synonym] = key
+
+    def load_from_config_file(self, filename):
+        parser = SafeConfigParser()
+        parser.read(filename)
+        data = {}
+        for section in parser.sections():
+            data.update(dict(parser.items(section)))
```

Comment:
Do we lose any backwards compatibility by flattening the storage here? Are there any cases currently where experiments might define a configuration parameter in a custom section using the same key as another parameter from some default section? This seems like potentially a big risk given that we don't know exactly how individual experiments may currently make use of configuration sections.